src/metax_api/services/file_service.py

```diff
@@ -210,78 +210,109 @@ def restore_files(cls, request, file_identifier_list):
-
-        return Response(preferred_identifiers, status=status.HTTP_200_OK)
-
-    @classmethod
-    def get_detailed_datasets_where_file_belongs_to(cls, file_identifiers):
-        """
-        Find out which (non-deprecated) datasets a list of files belongs to, and return
-        their preferred_identifiers per file as a list in json format.
-
-        Parameter file_identifiers can be a list of pk's (integers), or file identifiers (strings).
-        """
-        _logger.info('Retrieving detailed list of datasets where files belong to')
-
-        file_ids = cls._file_identifiers_to_ids(file_identifiers)
-
-        _logger.info('Looking datasets for the following files (printing first 10):\n%s'
-                     % '\n'.join(str(id) for id in file_identifiers[:10]))
-
-        sql_select_related_records = """
+        files = """
             SELECT f.identifier, json_agg(cr.research_dataset->>'preferred_identifier')
             FROM metax_api_file f
             JOIN metax_api_catalogrecord_files cr_f
                 ON f.id=cr_f.file_id
             JOIN metax_api_catalogrecord cr
                 ON cr.id=cr_f.catalogrecord_id
             WHERE f.id IN %s
-                AND cr.removed = false AND cr.active = true
+                AND cr.removed = false AND cr.active = true AND cr.deprecated = false
+            GROUP BY f.id
+            ORDER BY f.id ASC;
+            """
+
+        files_keysonly = """
+            SELECT f.identifier
+            FROM metax_api_file f
+            JOIN metax_api_catalogrecord_files cr_f
+                ON f.id=cr_f.file_id
+            JOIN metax_api_catalogrecord cr
+                ON cr.id=cr_f.catalogrecord_id
+            WHERE f.id IN %s
+                AND cr.removed = false AND cr.active = true AND cr.deprecated = false
             GROUP BY f.id
             ORDER BY f.id ASC;
             """
 
+        datasets_keysonly = """
+            SELECT cr.research_dataset->>'preferred_identifier' AS preferred_identifier
+            FROM metax_api_file f
+            JOIN metax_api_catalogrecord_files cr_f
+                ON f.id=cr_f.file_id
+            JOIN metax_api_catalogrecord cr
+                ON cr.id=cr_f.catalogrecord_id
+            WHERE cr.id IN %s
+                AND cr.removed = false AND cr.active = true AND cr.deprecated = false
+            GROUP BY cr.id
+            ORDER BY cr.id ASC;
+            """
+
+        datasets = """
+            SELECT cr.research_dataset->>'preferred_identifier', json_agg(f.identifier)
+            FROM metax_api_file f
+            JOIN metax_api_catalogrecord_files cr_f
+                ON f.id=cr_f.file_id
+            JOIN metax_api_catalogrecord cr
+                ON cr.id=cr_f.catalogrecord_id
+            WHERE cr.id IN %s
+                AND cr.removed = false AND cr.active = true AND cr.deprecated = false
+            GROUP BY cr.id
+            ORDER BY cr.id ASC;
+            """
+
```

Comment:
'files_keysonly' and 'datasets_keysonly' are pretty much copy paste sql code from 'files' and 'datasets'. Would it be nicer to parameterise the difference in input for those...?

Reply:
Good point, done with string.replace()